server/bin/pbench-index.py

```diff
@@ -181,393 +190,16 @@ def main(options, name):
-    tarballs = []
-    try:
-        tb_glob = os.path.join(ARCHIVE_rp, "*", linksrc, "*.tar.xz")
-        for tb in glob.iglob(tb_glob):
-            try:
-                rp = Path(tb).resolve(strict=True)
-            except OSError:
-                idxctx.logger.warning("{} does not resolve to a real path", tb)
-                quarantine(qdir, idxctx.logger, tb)
-                continue
-            controller_path = rp.parent
-            controller = controller_path.name
-            archive_path = controller_path.parent
-            if str(archive_path) != str(ARCHIVE_rp):
-                idxctx.logger.warning(
-                    "For tar ball {}, original home is not {}", tb, ARCHIVE_rp
-                )
-                quarantine(qdir, idxctx.logger, tb)
-                continue
-            if not Path(f"{rp}.md5").is_file():
-                idxctx.logger.warning("Missing .md5 file for {}", tb)
-                quarantine(qdir, idxctx.logger, tb)
-                # Audit should pick up missing .md5 file in ARCHIVE directory.
-                continue
-            try:
-                # get size
-                size = rp.stat().st_size
-            except OSError:
-                idxctx.logger.warning("Could not fetch tar ball size for {}", tb)
-                quarantine(qdir, idxctx.logger, tb)
-                # Audit should pick up missing .md5 file in ARCHIVE directory.
-                continue
-            else:
-                tarballs.append((size, controller, tb))
-    except Exception:
-        idxctx.logger.exception(
-            "Unexpected error encountered generating list" " of tar balls to process"
-        )
-        return 12
-    else:
-        if not tarballs:
-            idxctx.logger.info("No tar balls found that need processing")
-            return 0
-
-    # We always process the smallest tar balls first.
-    tarballs = sorted(tarballs)
+    idx = Index(name, options, idxctx, INCOMING_rp)
+    res = idx.collect_tb(ARCHIVE_rp, qdir)
 
-    # At this point, tarballs contains a list of tar balls sorted by size
-    # that were available as symlinks in the various 'linksrc' directories.
-    idxctx.logger.debug("Preparing to index {:d} tar balls", len(tarballs))
-
-    try:
-        # Now that we are ready to begin the actual indexing step, ensure we
-        # have the proper index templates in place.
-        idxctx.logger.debug("update_templates [start]")
-        idxctx.templates.update_templates(idxctx.es)
-    except TemplateError as e:
-        idxctx.logger.error("update_templates [end], error {}", repr(e))
-        res = 9
-    except Exception:
-        idxctx.logger.exception(
-            "update_templates [end]: Unexpected template" " processing error"
-        )
-        res = 12
-    else:
-        idxctx.logger.debug("update_templates [end]")
-        res = 0
-
-    if res != 0:
-        # Exit early if we encounter any errors.
+    if isinstance(res, int):
```

Comment:
Overloading the return value really isn't a good strategy for errors. I think we should probably raise exceptions that are handled by the driver code (here) and decoded to defined exit codes if that's really necessary.

In retrospect, I wonder if the distinct exit codes (aside from the 0/success or nonzero/failure meaningful to the shell) really serve any useful purpose.

We don't need to fix that completely in this PR, but I would like to fix the return value overloading we have here. One simple alternative would be to return a tuple of `(status, tarball_list)` which would be for example `(2, [])` on a failure or `(0, [...])` on success.

Reply:
+1 to using a tuple return value.

With regard to status code values, we used those codes to debug why the Tool Meister was not running problem in an environment, so they can be very useful.

Reply:
Distinguishing various error conditions is certainly important, but relying on log messages (particularly if they have useful parameters) seems a lot better than trying to backtrack and guess from integer return values without additional context. And if we have log messages, seems to me that the individual integer codes are at best redundant?

Reply:
Yes, until you don't have logs from the system where the code ran.  =)